src/test/shell/bazel/target_compatible_with_test.sh

```diff
@@ -0,0 +1,301 @@
+  expect_log '^//:fail_on_foo2  *  FAILED in'
+  expect_log '^//:pass_on_foo1_bar2  *  SKIPPED$'
+
+  # Use :all for this one to validate similar behaviour.
+  if ! bazel test --host_platform=@//:foo1_bar2_platform \
+    --platforms=@//:foo1_bar2_platform :all &> "${TEST_log}"; then
+    cat "${TEST_log}"
+    fail "Bazel failed unexpectedly."
+  fi
+  expect_log 'Executed 2 out of 3 tests: 2 tests pass and 1 was skipped'
+  expect_log '^//:pass_on_foo1  *  PASSED in'
+  expect_log '^//:fail_on_foo2  *  SKIPPED$'
+  expect_log '^//:pass_on_foo1_bar2  *  PASSED in'
+}
+
+function test_failure_on_incompatible_top_level_target() {
+  cd target_skipping || fail "couldn't cd into workspace"
+
+  if bazel test --show_result=10 --host_platform=@//:foo1_bar1_platform \
+    --platforms=@//:foo1_bar1_platform //:pass_on_foo1_bar2 \
+    --build_event_text_file="${TEST_log}".build.json \
+    &> "${TEST_log}"; then
+    fail "Bazel passed unexpectedly."
+  fi
+
+  expect_log '^Target //:pass_on_foo1_bar2 failed to build'
+  expect_log '^ERROR: Some incompatible targets were explicitly requested$'
+  expect_log '^FAILED: Build did NOT complete successfully'
+  expect_log '^//:pass_on_foo1_bar2  *  NO STATUS$'
+
+  # Now look at the build event log.
+  mv "${TEST_log}".build.json "${TEST_log}"
+
+  expect_log "^  reason: ANALYSIS_FAILURE$"
+  expect_log "Target //:pass_on_foo1_bar2 is incompatible and cannot be built."
+
+  # Run an additional (passing) test and make sure we still fail the build.
+  if bazel test --show_result=10 --host_platform=@//:foo1_bar1_platform \
+    --platforms=@//:foo1_bar1_platform //:pass_on_foo1_bar2 //:pass_on_foo1 \
+    &> "${TEST_log}"; then
+    fail "Bazel passed unexpectedly."
+  fi
+
+  expect_log '^//:pass_on_foo1  *  PASSED in'
+  expect_log '^ERROR: Some incompatible targets were explicitly requested$'
+  expect_log '^FAILED: Build did NOT complete successfully'
+}
+
+function test_build_event_protocol() {
+  cd target_skipping || fail "couldn't cd into workspace"
+
+  if ! bazel test --show_result=10 --host_platform=@//:foo1_bar1_platform \
+    --build_event_text_file=$TEST_log \
+    --platforms=@//:foo1_bar1_platform ...; then
+    cat "${TEST_log}"
+    fail "Bazel failed unexpectedly."
+  fi
+  expect_not_log 'Target //:pass_on_foo1 build was skipped\.'
+  expect_log_n 'reason: SKIPPED\>' 3
+  expect_log 'Target //:fail_on_foo2 build was skipped\.'
+  expect_log 'Target //:pass_on_foo1_bar2 build was skipped\.'
+}
+
+function test_non_top_level_skipping() {
+  # TODO(phil): We still build dependencies even if they're incompatible. This
+  # needs to be fixed in a future patch.
+  cat >> target_skipping/BUILD <<EOT
+genrule(
+    name = "genrule_foo1",
+    target_compatible_with = [":foo1"],
+    outs = ["foo1.sh"],
+    cmd = "echo 'Should not be executed' &>2; exit 1",
+)
+
+sh_binary(
+    name = "sh_foo2",
+    srcs = ["foo1.sh"],
+    target_compatible_with = [":foo2"],
+)
+EOT
```

Comment:
All the tests in this file pass except fro this one.

This test makes bazel unhappy in the following way:
```
Analyzing: target //:sh_foo2 (1 packages loaded, 0 targets configured)
Internal error thrown during build. Printing stack trace: java.lang.RuntimeException: Unrecoverable error while evaluating node '//:genrule_foo1 BuildConfigurationValue.Key[5f35dd882ad57447c6b7d36e650cd887e37a43ccf9f72cef887e8f5c4fd12858]' (requested by nodes '//:foo1.sh BuildConfigurationValue.Key[5f35dd882ad57447c6b7d36e650cd887e37a43ccf9f72cef887e8f5c4fd12858]')                                 
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:513)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:398)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.IllegalStateException: genrule rule //:genrule_foo1 : These artifacts do not have a generating action:
bazel-out/k8-fastbuild/bin/foo1.sh
No origin, run with --experimental_extended_sanity_checks
These actions were checked:
com.google.devtools.build.lib.analysis.actions.SymlinkAction Symlink
    bazel-out/k8-fastbuild/bin/genrule_foo1
com.google.devtools.build.lib.analysis.SourceManifestAction SourceSymlinkManifest
    bazel-out/k8-fastbuild/bin/genrule_foo1.runfiles_manifest
com.google.devtools.build.lib.analysis.actions.SymlinkTreeAction SymlinkTree
    bazel-out/k8-fastbuild/bin/genrule_foo1.runfiles/MANIFEST
com.google.devtools.build.lib.actions.MiddlemanAction Middleman
    bazel-out/k8-fastbuild/internal/_middlemen/genrule_Ufoo1-runfiles

        at com.google.devtools.build.lib.analysis.CachingAnalysisEnvironment.verifyGeneratedArtifactHaveActions(CachingAnalysisEnvironment.java:156)
        at com.google.devtools.build.lib.analysis.CachingAnalysisEnvironment.disable(CachingAnalysisEnvironment.java:119)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.createConfiguredTarget(ConfiguredTargetFunction.java:1015)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.compute(ConfiguredTargetFunction.java:354)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:438)
        ... 7 more
```
I have tried a few things (like teaching `CachingAnalysisEnvironment.disable()` about incompatible platform skipping, but kept running into other issues.

Somewhere I have to teach bazel about skipped dependencies, but haven't figured out where yet. Will resume on this next week.

Reply:
Yeah, adding awareness of this to `CachingAnalysisEnvironment.disable()` was my thought too. What other issues came up?